File: Algorithms/Java/Valid Number.java
public class Solution {
    public boolean isNumber(String s) {
    	boolean num = false;
		boolean dot = false;
		boolean exp = false;
		String scut = s.trim();
		int length = scut.length();
		for (int i = 0; i < length; i++) {
			char c = scut.charAt(i);
			if (c == '.') {
				if (exp || dot) {
					return false;
				}
				dot = true;
			} else if (c <= '9' && c >= '0')
				num = true;
			else if (c == '+' || c == '-') {
				if (i != 0 && scut.charAt(i - 1) != 'e')
					return false;
			} else if (c == 'e') {
				if ((!num) || exp) {
					return false;
				}
				num = false; // "2e2" is valid but "2e" is not.
				exp = true;
			} else
				return false;
		}
		return num;
    }
}

File: Algorithms/Python/Valid-Number.py
class Solution(object):

    def isNumber(self, s):
        """
        :type s: str
        :rtype: bool
        """
        s = s.strip()
        if not s:
            return False
        num, exp, dot = False, False, False
        for i, x in enumerate(s):
            if x.isdigit():
                num = True
            elif x in '+-':
                if i > 0 and s[i - 1] != 'e':
                    return False
            elif x == '.':
                if dot or exp:
                    return False
                dot = True
            elif x == 'e':
                if exp or not num:
                    return False
                exp = True
                num = False
            else:
                return False
        return num
